src/Mvc/Mvc.Core/src/Routing/ControllerActionEndpointDataSourceIdProvider.cs

```diff
@@ -0,0 +1,17 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    internal class ControllerActionEndpointDataSourceIdProvider
```

Comment:
Is there the opportunity to share this infrastructure between controllers and pages? This is located in Mvc.Core and I believe there is an InternalsVisibleTo share to the razor pages assembly.

I'm not a fan of sharing internal code, but this is within the framework, and I'm sure many other internal shares already.

Reply:
I think DynamicControllerEndpointSelectorCache could be shared by changing it to work with base types. Or made generic.

Reply:
There's probably something that can be done, but do you think it's worth it?

I want to have two different types for the ids since that makes it impossible to mix them between controllers and actions. This is a common pattern that we follow across MVC for Controllers and Pages.

I would agree if the code was more complex, but this is all plumbing, so I don't think there's a lot of value in removing duplication.

That said, If you feel strongly about it, let me know and I'll do it.

Reply:
> I want to have two different types for the ids since that makes it impossible to mix them between controllers and actions. This is a common pattern that we follow across MVC for Controllers and Pages.

Why would that be bad? The ID is an incrementing value that has no meaning other than being unique identifier.

If there was more time I'd like to explore reducing duplication, but I don't feel strongly about it.

Reply:
Fair enough, I'll take a look on monday and consider a fix on master (6.0)